File: 프로그래머스/lv2/17680. ［1차］ 캐시/［1차］ 캐시.cpp
#include <bits/stdc++.h>
using namespace std;

string lower(string a){
    string ret;
    for(auto i : a) ret += tolower(i);
    return ret; 
}

int solution(int cacheSize, vector<string> cities) {
    vector<string> cache;
    int ret = 0;
    
    for(auto city: cities){
        string lo = lower(city);
        auto it = find(cache.begin(), cache.end(), lo);
        
        if(it == cache.end()){
            ret += 5;
            cache.push_back(lo);
            
            if(cacheSize < cache.size()){
                cache.erase(cache.begin());
            }
        } else{
            ret += 1;
            cache.erase(it);
            cache.push_back(lo);            
        }        
    }    
    return ret;
}

File: 프로그래머스/lv2/17680. ［1차］ 캐시/［1차］ 캐시.java
import java.util.*;

class Solution {
    public int solution(int cacheSize, String[] cities) {
        
        int ret = 0;
        ArrayList cache = new ArrayList();
        
        for(int i = 0; i < cities.length; i++){
            String lo = cities[i].toLowerCase();
            
            if(cache.remove(lo)){
                ret += 1;
                cache.add(lo);
            } else {
                ret += 5;
                cache.add(lo);
                
                if(cacheSize < cache.size()){
                    cache.remove(0);
                }
            }            
        }
        return ret;
    }
    
    
    
}